src/Compilers/CSharp/Portable/Binder/Binder_Symbols.cs

```diff
@@ -1661,21 +1661,40 @@ internal Symbol GetSpecialTypeMember(SpecialMember member, BindingDiagnosticBag
                 : null;
         }
 
-        internal static bool TryGetSpecialTypeMember<TSymbol>(CSharpCompilation compilation, SpecialMember specialMember, SyntaxNode syntax, BindingDiagnosticBag diagnostics, out TSymbol symbol)
+        internal static bool TryGetSpecialTypeMember<TSymbol>(CSharpCompilation compilation, SpecialMember specialMember, SyntaxNode syntax, BindingDiagnosticBag diagnostics, out TSymbol symbol, bool isOptional = false)
             where TSymbol : Symbol
         {
             symbol = (TSymbol)compilation.GetSpecialTypeMember(specialMember);
             if (symbol is null)
             {
-                MemberDescriptor descriptor = SpecialMembers.GetDescriptor(specialMember);
-                diagnostics.Add(ErrorCode.ERR_MissingPredefinedMember, syntax.Location, descriptor.DeclaringTypeMetadataName, descriptor.Name);
+                if (!isOptional)
+                {
+                    MemberDescriptor descriptor = SpecialMembers.GetDescriptor(specialMember);
+                    diagnostics.Add(ErrorCode.ERR_MissingPredefinedMember, syntax.Location, descriptor.DeclaringTypeMetadataName, descriptor.Name);
+                }
+
                 return false;
             }
 
             var useSiteInfo = GetUseSiteInfoForWellKnownMemberOrContainingType(symbol);
             if (useSiteInfo.DiagnosticInfo != null)
             {
-                diagnostics.ReportUseSiteDiagnostic(useSiteInfo.DiagnosticInfo, new SourceLocation(syntax));
+                // Report errors only for non-optional members
+                if (isOptional)
+                {
+                    var severity = useSiteInfo.DiagnosticInfo.Severity;
+
+                    // If the member is optional and bad for whatever reason ignore it
+                    if (severity == DiagnosticSeverity.Error)
```

Comment:
Why don't we want to report warnings in this case? It seems likely to me that this will cause missed use site diagnostics/tracking somewhere.

Reply:
This is in sync with existing logic in `GetWellKnownTypeMember`, wich also swallows use site warnings in case the requested member is optional:
https://github.com/dotnet/roslyn/blob/30e683e2a6953a446a34ac5fc499287664061b96/src/Compilers/CSharp/Portable/Binder/Binder_Symbols.cs#L1805-L1818
TBH I don't fully understand what use site diagnostics even are, but it is intuitive that similar concepts should have similar behavior. And this change was suggested by Aleksey

Reply:
> Why don't we want to report warnings in this case?

I think the reason is to avoid an unnecessary noise for methods that user didn't use explicitly. This is a preexisting behavior for well-known members.

>  It seems likely to me that this will cause missed use site diagnostics/tracking somewhere.

I cannot think of any reason why this would be the case. Do you have something specific in mind?

Reply:
> I cannot think of any reason why this would be the case. Do you have something specific in mind?

I am unsure whether we have any use-site diagnostics today that are just warnings. However, if we do, and if this method has such a warning, it's not going to be reported by the LocalRewriter machinery that creates the synthesized `BoundCall` here, and (though this is unlikely to matter in reality, especially given that this a SpecialType and not a WellKnownType) the use of the assembly symbol won't be tracked from this call.

> I think the reason is to avoid an unnecessary noise for methods that user didn't use explicitly

This argument doesn't make sense to me. Regardless of whether they intended to use it, they are using it, and it can have diagnostics. The pre-existing behavior for well-known members sounds like a bug.

Reply:
> if we do, and if this method has such a warning, it's not going to be reported by the LocalRewriter machinery that creates the synthesized BoundCall

That is the intended behavior that we already have for well-known members.

> the use of the assembly symbol won't be tracked from this call

Why is that? The use of assembly symbols is not tracked based on errors or warnings. And, if you carefully review the code of this function, you will notice that nothing about tracking used references is changed in this method. They are tracked exactly the same way they used to be tracked before.

> The pre-existing behavior for well-known members sounds like a bug.

I am happy to have this discussion outside of context of this PR. For now, I think we should simply follow what we were doing for well-known members. So, I suggest to not block this PR based on this.

Reply:
I'm ok to have the discussion outside the PR.